nucliadb/src/nucliadb/common/cluster/rebalance.py

```diff
@@ -18,162 +18,444 @@
+
+    async def get_rebalance_shards(self) -> list[RebalanceShard]:
+        """
+        Return the sorted list of shards by increasing paragraph count.
+        """
+        self.kb_shards = await datamanagers.atomic.cluster.get_kb_shards(kbid=self.kbid)
+        if self.kb_shards is None:
+            return []
+        return list(
+            sorted(
+                [
+                    RebalanceShard(
+                        id=shard.shard,
+                        nidx_id=shard.nidx_shard_id,
+                        paragraphs=await get_shard_paragraph_count(shard.nidx_shard_id),
+                        active=(idx == self.kb_shards.actual),
+                    )
+                    for idx, shard in enumerate(self.kb_shards.shards)
+                ],
+                key=lambda x: x.paragraphs,
+            )
         )
-        results[shard_meta.shard] = shard_data.paragraphs
 
-    return [(shard, paragraphs) for shard, paragraphs in sorted(results.items(), key=lambda x: x[1])]
+    async def build_shard_resources_index(self):
+        rids = [rid async for rid in datamanagers.resources.iterate_resource_ids(kbid=self.kbid)]
+        async with datamanagers.with_ro_transaction() as txn:
+            for resources_batch in batchify(rids, n=100):
+                shards = await txn.batch_get(
+                    keys=[KB_RESOURCE_SHARD.format(kbid=self.kbid, uuid=rid) for rid in resources_batch],
+                    for_update=False,
+                )
+                for rid, shard_bytes in zip(resources_batch, shards):
+                    if shard_bytes is not None:
+                        self.index.setdefault(shard_bytes.decode(), set()).add(rid)
+
+    async def move_paragraphs(
+        self, from_shard: RebalanceShard, to_shard: RebalanceShard, max_paragraphs: int
+    ) -> int:
+        """
+        Takes random resources from the source shard and tries to move at most max_paragraphs.
+        It stops moving paragraphs until the are no more resources to move.
+        """
+        moved_paragraphs = 0
+
+        while moved_paragraphs < max_paragraphs:
+            # Take a random resource to move
+            try:
+                resource_id = random.choice(tuple(self.index[from_shard.id]))
+            except (KeyError, IndexError):
+                # No more resources in shard or shard not found
+                break
+
+            assert self.kb_shards is not None
+            from_shard_obj = next(s for s in self.kb_shards.shards if s.shard == from_shard.id)
+            to_shard_obj = next(s for s in self.kb_shards.shards if s.shard == to_shard.id)
+            paragraphs_count = await get_resource_paragraphs_count(resource_id, from_shard.nidx_id)
+            moved = await move_resource_to_shard(
+                self.context, self.kbid, resource_id, from_shard_obj, to_shard_obj
+            )
+            if moved:
+                self.index[from_shard.id].remove(resource_id)
```

Comment:
Maybe the resource should be added to the index of the destination shard here?

Still reviewing, not sure if this is relevant.